### ISSUE TITLE
catch(Exception) does not re-throw

### ISSUE THREAD

**Issue by jinek** [DataDog/dd-trace-dotnet#773]
In catch block https://github.com/DataDog/dd-trace-dotnet/blob/388fa1ceb60e2ae4bd01355d0fa46320f1b73469/src/Datadog.Trace.ClrProfiler.Managed/Integrations/WcfIntegration.cs#L156
there is no re-throw of the exception.
Thus the exception does not propogate to the caller and, basically, infrastructure does not know when logging failed.
Which means WCF servers requests and they can't be seen from the datadog.

For instance, if **OutOfMemoryException** happened inside try block, host will not know about the problem.

**Comment by macrogreg** [comment]
Hi @jinek ,

Lots of great submissions. Thank you!
What is the specific scenario you are trying to achieve?

In general, instrumentation should never crash an application. So if an exception happened within the instrumentation code (not the application code being instrumented), we want to log it for the customer to see and then we want to proceed.

Given the choice between failing to create a crate and crashing the app, we will always choose failing to create a trace. Of course, ideally neither should happen. :)

**Comment by macrogreg** [comment]
Hi @jinek 

I have not heard back, so I will close the issue for now.
However, please feel free to re-open if you have new information.

Thanks!

**Comment by macrogreg** [comment]
Hi @jinek ,

One of our most important requirements is that an application that did not crash without the Tracer, must continue not crashing with the Tracer. We cannot know if the customer waits for the exception and so if some memory issue or similar occurred due to tracer code (like in your example), then we will always do our best to recover and back-off gracefully rather than to let the exception escape. 

For exceptions that happened in application code for application specific reasons, it is, of course different. There, we do what you expect: log the application exception and let it propagate.

We will take the "exception event" that you describe into a design discussion. 
In the meantime you can use an AddDomain exception handler as a workaround:
https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.firstchanceexception

Please let me know if you have more questions. Thanks!

**Comment by jinek** [comment]
I understand your point.

Just consider, please, my opinion:

> We cannot know if the customer waits for the exception

It's same for any public library/tool developer: nobody knows whether customer waits for exception or no. Still, this should be the decision of the customer - implement own handling or to allow default .NET runtime handler to close the application, set the result code to 1, log to windows etc. 

> do our best to recover

It's not possible to recover from the state which you don't know about. This is the point of basic c# analysis rules:
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1031?view=vs-2019

In the example of OutOfMemoryException, imagine that host of application is waiting for OutOfMemoryException to re-run the application with bigger memory set.  Or may be it is waiting for AssemblyLoadException to re-deploy the application.

> Requirements is that an application that did not crash without the Tracer, must continue not crashing with the Tracer

This just seems to me a strange requirment. If I call a library or a tool, I expect it either to work or to return an error.
In case of error I can switch it off, report the issue, wait for the fix and try one more time.
But if it continues to fail silently, it is only performance degradation, which I don't see, nothing been reported to me etc.

> For exceptions that happened in application code for application specific reasons

The code of dd-tracer make lot of calls to the application and it's runtime. My point is inside your catch you can not know if that exception comes from application code or no.

Out of topic: while looking at the code, I've started to suspect an issue, have a look at the flow please:

In the mention code of dd-tracer, the method **CreateScope** returns null if it catches an exception or RequestContext is of another type. Which means that method **HandleRequest** will throw NullReferenceException at the line scope.Span.SetException(ex);. 
This way, instead of having the original exception from HttpContext (which happened first and customer may be is expecting to get. Or may be it's HttpRequestException which is awaited by aspnet) we are receiving third one - NullReferenceException.

**Comment by macrogreg** [comment]
When somebody makes a library call they will make a decision on whether or not to catch exceptions and how to handle them.

On contrary, a tracer is often attached to an application that is already in production and will not be modified. That is a critical difference.

This is a great discussion, but it turned theoretical. Is it more about philosophy of error handling, or is there a concrete scenario you are trying to achieve that is not working? I am not sure how an Out Of Memory exception would be handled, but perhaps you can clarify? :)

**Comment by jinek** [comment]
I have few scenarios, but that is not in priority. I will report to you for sure on its time.

**OutOfMemoryException** is same as other exceptions. For example **NullReferenceException** is full analogy.
But I don't understand the question, could you rephrase please?

I'm talking about **OutOfMemoryException** as an example of the fact that runtime can throw exceptions by itself independently of what we are doing. In fact there are few dozens of them and some of them can be thrown from empty line of code in empty method.  If we wrap entire method in to try/catch, it still can be raised before that try.

Not sure if you aware of this and if it could help you, but additionally there are several special cases: **ThreadAbortException** is always re-thrown automatically at the end of a catch and is always automatically caught by managed threads. **StackOverflowException** and some others can't be caught at all and always reaches the default runtime handler (which can be overridden tho).

**Comment by macrogreg** [comment]
> I have few scenarios, but that is not in priority. I will report to you for sure on its time.

Ok, sounds good, @jinek .

**Comment by jinek** [comment]
@macrogreg  Hello again!
May be this is somehow related issue: if we cancel the request on client side, in datadog we have this marked as failure:
![image](https://user-images.githubusercontent.com/10516222/98036539-80928100-1e1a-11eb-9782-f1110ab04022.png)
